frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts

```diff
@@ -96,7 +101,38 @@ export function getWaterfallDataset(
     dataset.push(createDatum({ dimension, barOffset, increase, decrease }));
   });
 
-  // TODO total
+  if (!settings["waterfall.show_total"]) {
+    return dataset;
+  }
+
+  const barOffset =
+    total >= 0 ? negativeTranslation : negativeTranslation + total;
+
+  let dimension = "Total";
+  // For timeseries x-axis ECharts will not allow mixed values,
+  // so we cannot set the dimension value to "Total." As a workaround,
+  // we instead set it to be a date after the final date in the dataset,
+  // then in the x-axis label formatter we will replace the label with the
+  // string "Total."
+  if (settings["graph.x_axis.scale"] === "timeseries") {
+    const lastDimensionValue = rows[rows.length - 1][columns.dimension.index];
+    if (typeof lastDimensionValue === "boolean") {
+      throw Error(
+        "dimension value cannot be boolean with timeseries x-axis scale",
+      );
+    }
+
+    const lastDate = dayjs(lastDimensionValue);
+    dimension = lastDate.add(1, "day").toISOString();
```

Comment:
I think it should not be 1 day but instead for structured aka GUI questions it should be `dimensionColumn.unit` but for native queries we need to compute the longest unit from data. Like if every row is the first day of month, then the unit is month and not week or day

This is how it looks on a GUI question with `year` unit
<img width="610" alt="Screenshot 2024-01-04 at 9 29 55 PM" src="https://github.com/metabase/metabase/assets/14301985/8acd8661-e239-4d16-93db-8b068ae42edd">